### ISSUE TITLE
Can't find monitor when running two displays in mirrored mode

### ISSUE THREAD

**Issue by jbaiter** [polybar/polybar#1191]
When I run an external screen in mirrored mode (i.e. the same viewport as my main monitor), polybar no longer recognizes my main monitor and exits with:
```
◆ polybar -r example
error: Monitor "eDP-1" not found or disconnected
```

Here is the relevant part of my config:
```
[default]
monitor = ${env:MONITOR:eDP-1}
```

XRandR still lists the monitor with the exact same name:

```
◆ xrandr --listmonitors
Monitors: 2
 0: +*eDP-1 2560/309x1440/174+0+0  eDP-1
 1: +HDMI-2 2560/677x1440/381+0+0  HDMI-2

◆ xrandr -q
Screen 0: minimum 320 x 200, current 2560 x 1440, maximum 8192 x 8192
eDP-1 connected primary 2560x1440+0+0 (normal left inverted right x axis y axis) 309mm x 174mm
   2560x1440     60.00*+
   // ....
HDMI-2 connected 2560x1440+0+0 (normal left inverted right x axis y axis) 0mm x 0mm
   1920x1080     60.00*+  50.00    59.94    30.00    25.00    24.00    29.97    23.98
   // ...
```

... but polybar does not recognize it:
```
◆ polybar --list-monitors
HDMI-2: 2560x1440+0+0
```

I am running the latest Git master with `WITH_XRANDR_MONITORS` enabled:
```
◆ polybar -v
polybar 3.1.0-110-gff39c9d-git

Features: -alsa +curl -i3 -mpd +network +pulseaudio
```

**Comment by patrick96** [comment]
What happens when you set the monitor as `HDMI-2`?

**Comment by jbaiter** [comment]
That works, but HDMI-2 is not always connected, it's a laptop and only occasionally connected to external displays for presentation purposes.
I dug into the code a bit, and commenting out the "clone removal" part in `randr_util::get_monitors` resolves the issue. The reasons seems to be that it has not way to detect which monitor clones which, so the removal is based on iteration order, which in this case removes the "main" display.

**Comment by patrick96** [comment]
Closing due to inactivity

**Comment by BVollmerhaus** [comment]
I'm running into the same issue; when cloning outputs, Polybar does not recognize them correctly:

```bash
$ polybar --list-monitors
DP-1-1: 1920x1080+0+0
```


```bash
$ xrandr --query | grep connected
LVDS1 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 340mm x 190mm
VGA1 disconnected (normal left inverted right x axis y axis)
... more disconnected screens...
DP-1-1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 518mm x 324mm
```

My script for launching bars on my primary and my secondary, only occasionally connected, monitors:

```bash
for screen in $(xrandr --query | grep -w connected); do
  output=${screen%% *}

  case ${screen} in
    *primary*)
      launch_bar "${output}" primary
      ;;
    *)
      launch_bar "${output}" secondary
      ;;
  esac
done
```

This works fine when I place the connected and internal monitor(s) side-by-side. The primary bar starts on my primary (internal) monitor and the secondary one starts on all other connected ones. However, when I clone the displays, i.e. place them on top of each other in `arandr`, Polybar only recognizes one of them:

```
error: Monitor "LVDS1" not found or disconnected
```

So I think this issue should be re-opened. I'd gladly assist in debugging this problem.

**Comment by Vincevrp** [comment]
I'm experiencing the exact same problem. 
```
xrandr --output eDP1 --primary --mode 1920x1080 --pos 0x1128 --rotate normal --output HDMI1 --mode 1920x1080 --pos 0x0 --rotate normal --same-as eDP1
```

When I run the command above to mirror my displays, polybar only recognizes HDMI1.

```
polybar -m

HDMI1: 1920x1080+0+0
```

xrandr output:

```
Screen 0: minimum 8 x 8, current 1920 x 1080, maximum 32767 x 32767
eDP1 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 350mm x 190mm
   1920x1080     59.93*+  59.93
   1680x1050     59.88
   1400x1050     59.98
   1600x900      60.00    59.95    59.82
   1280x1024     60.02
   1400x900      59.96    59.88
   1280x960      60.00
   1368x768      60.00    59.88    59.85
   1280x800      59.81    59.91
   1280x720      59.86    60.00    59.74
   1024x768      60.00
   1024x576      60.00    59.90    59.82
   960x540       60.00    59.63    59.82
   800x600       60.32    56.25
   864x486       60.00    59.92    59.57
   640x480       59.94
   720x405       59.51    60.00    58.99
   640x360       59.84    59.32    60.00
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
HDMI1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 530mm x 300mm
   1920x1080     60.00*+  50.00    59.94
   1920x1080i    60.00    50.00    59.94
   1600x1200     60.00
   1600x900      60.00
   1280x1024     75.02    60.02
   1152x864      75.00
   1280x720      60.00    50.00    59.94
   1024x768      75.03    60.00
   800x600       75.00    60.32
   720x576       50.00
   720x576i      50.00
   720x480       60.00    59.94
   720x480i      60.00    59.94
   640x480       75.00    60.00    59.94
   720x400       70.08
HDMI2 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

**Comment by Yves0** [comment]
I'm experiencing the problem too.
I sets two monitor to mirror mode `xrandr --output eDP1  --output HDMI1 --same-as eDP1`
then start polybar with script
```
#!/usr/bin/env bash

# Terminate already running bar instances
killall -q polybar

# Wait until the processes have been shut down
while pgrep -u $UID -x polybar >/dev/null; do sleep 1; done

# Launch bar1 and bar2
#polybar top
if type "xrandr"; then
  for m in $(xrandr --query | grep " connected" | cut -d" " -f1); do
    MONITOR=$m polybar --reload top &
  done
else
  polybar --reload top &
fi

echo "Bars launched..."
```
here is the error log
```
error: Monitor "eDP1" not found or disconnected
```

But anyway, polybar runs with the another monitor. HOWEVER, i3 module disappeared.
After setting `pin-workspaces = false`, i3 module runs normally. But every time I change the screen layout,I have to change the polybar config file too.

**Comment by patrick96** [comment]
As @jbaiter correctly pointed out, for some reason polybar purges monitors that appear to be mirrored from the list of monitor it recognizes. I have not found the reason why this is done or what would break if we remove that part.

From some quick testing, just removing the part that removes cloned monitors does indeed fix this particular issue. I used the following patch:

```patch
diff --git a/src/x11/extensions/randr.cpp b/src/x11/extensions/randr.cpp
index 36baedb..57d3192 100644
--- a/src/x11/extensions/randr.cpp
+++ b/src/x11/extensions/randr.cpp
@@ -141,34 +141,14 @@ namespace randr_util {
       }
     }
 
-    // clang-format off
-    const auto remove_monitor = [&](const monitor_t& monitor) {
-      monitors.erase(find(monitors.begin(), monitors.end(), monitor));
-    };
-    // clang-format on
-
-    for (auto m = monitors.rbegin(); m != monitors.rend(); m++) {
-      if ((*m)->w == 0) {
-        remove_monitor(*m);
-        continue;
+    // Remove empty monitors
+    auto it = monitors.begin();
+    while (it != monitors.end()) {
+      if((*it)->w == 0) {
+        it = monitors.erase(it);
       }
-
-      // Test if there are any clones in the set
-      for (auto& monitor : monitors) {
-        if ((*m) == monitor || monitor->w == 0) {
-          continue;
-        } else if (check_monitor_support() && (monitor->output == XCB_NONE || (*m)->output == XCB_NONE)) {
-          continue;
-        }
-
-        // clang-format off
-        if (monitor->x >= (*m)->x && monitor->x + monitor->w <= (*m)->x + (*m)->w &&
-            monitor->y >= (*m)->y && monitor->y + monitor->h <= (*m)->y + (*m)->h) {
-          // Reset width so that the output gets
-          // removed in the base loop
-          monitor->w = 0;
-        }
-        // clang-format on
+      else {
+        it++;
       }
     }
 
```

I will need some more testing related to other parts of polybar that use the list of monitors (mainly the workspace modules) before I actually make those changes.

The one thing this removal of mirrored monitors could be useful for is if you iterate over `polybar -m` and launch a bar for each monitor listed. This behavior should probably be preserved. Maybe we could introduce an extra flag to `-m` that will print all monitors while `polybar -m` still removes mirrored monitors.

**Comment by tareefdev** [comment]
Me too have exactly the same problem. When the external monitor mirrors the internal one; Polybar crash with:
````
error: Monitor "eDP1" not found or disconnected
````

**Comment by Phil-Ba** [comment]
Having the same issue. Kind of annoying when I want to present something by mirroring my laptop display but would want to use the nm-applet to change the wifi.

**Comment by mjhough** [comment]
I am also having this issue. When I plug in an HDMI cable into my laptop it automatically mirrors the displays, causing polybar to crash. As mentioned above, when mirroring `polybar --list-monitors` lists only one monitor, the external one.

**Comment by patrick96** [comment]
I have written a possible fix for this in #1823, please check it out and see if it also fixes it for you.

**Comment by Zayon** [comment]
Sharing my workaround for presentation mode (mirrored screens)

```bash
#!/usr/bin/env sh

# Terminate already running bar instances
killall -q polybar

# Wait until the processes have been shut down
while pgrep -x polybar >/dev/null; do sleep 1; done

screens=$(xrandr --listactivemonitors | grep -v "Monitors" | cut -d" " -f6)

if [[ $(xrandr --listactivemonitors | grep -v "Monitors" | cut -d" " -f4 | cut -d"+" -f2- | uniq | wc -l) == 1 ]]; then
  MONITOR=$(polybar --list-monitors | cut -d":" -f1) TRAY_POS=right polybar main &
else
  primary=$(xrandr --query | grep primary | cut -d" " -f1)

  for m in $screens; do
    if [[ $primary == $m ]]; then
        MONITOR=$m TRAY_POS=right polybar main &
    else
        MONITOR=$m TRAY_POS=none polybar secondary &
    fi
  done
fi
```

With

```ini
[section/base]
monitor = ${env:MONITOR:eDP1}
```

`xrandr --listactivemonitors | grep -v "Monitors" | cut -d" " -f4 | cut -d"+" -f2- | uniq | wc -l` explained: 

`xrandr --listactivemonitors`: list active monitors :scream: 

```
Monitors: 2
 0: +*eDP1 1920/340x1080/190+0+0  eDP1
 1: +DP3 1920/530x1080/300+0+0  DP3
```

`grep -v "Monitors"`: removes the first line that is not useful

```
 0: +*eDP1 1920/340x1080/190+0+0  eDP1
 1: +DP3 1920/530x1080/300+0+0  DP3
```

` cut -d" " -f4`: keep only the resolution & position part of the output

```
1920/340x1080/190+0+0
1920/530x1080/300+0+0
```

` cut -d"+" -f2-`: keep only the position x+y part of the output

```
0+0
0+0
```

`uniq | wc -l`: removes duplicates then count the lines, if only 1 line remains then all screen are at the same position (mirrored)

`MONITOR=$(polybar --list-monitors | cut -d":" -f1) polybar main &` makes sure we launch polybar with the minitor that polybar will accept

Hope this helps
